Hcsvlab 1122

shaun-intersect opened the pull request.

dyazbek-intersect commented on db/seed_helper.rb:
```diff
@@ -5,4 +5,15 @@ def create_roles_and_permissions
-
+# Populates the languages table with the language names and codes from the languages CSV file
+def populate_languages
+  Language.delete_all
```
If you have an association between one model and a Language, and you delete all languages when the application is redeployed, then you have effectively broken that link.
You can use this: http://apidock.com/rails/v4.0.2/ActiveRecord/Relation/find_or_create_by

shaun-intersect replied:
The Language is essentially just being used as a storage/container for populating a dropdown and doesn't have any associations with other models. Languages would also only be created from the CSV instead of from user input/interaction, so there shouldn't be any data loss by deleting all of them (since it is re-imported from the CSV straight after deletion).

dyazbek-intersect replied:
Yes, for now this is true, but what if somebody decides to create associations to a Language from other models? You would save hours of debugging an issue that would show in production by fixing this now - it's a 1 line change.

dyazbek-intersect commented on db/seed_helper.rb:
```diff
@@ -5,4 +5,15 @@ def create_roles_and_permissions
+# Populates the languages table with the language names and codes from the languages CSV file
+def populate_languages
+  Language.delete_all
+  languages = {}
```
No point having an empty hash, then copying what you want out of the CSV then iterating across it. Just construct Models as you are iterating across the CSV.

dyazbek-intersect commented on db/seeds.rb:
```diff
@@ -8,3 +8,4 @@
 require File.dirname(__FILE__) + '/seed_helper.rb'
 
 create_roles_and_permissions
+populate_languages
```
Is this called each time you deploy? If not, you will need to modify capistrano config to get your new task being executed.

shaun-intersect replied:
I don't understand the system enough to confirm that, but I think it is based on the discussion I had with Ilya.

lotophage replied:
Looking at the cap tasks, it looks like this will only be called when the cap seed task is called. This also deletes any existing roles/languages, so perhaps instead it should check to see if the corresponding role/language record already exist in the DB and create it if it doesn't (otherwise it's going to bork any existing relations)? We could then add the seed task as a subtask of the deploy task.

dyazbek-intersect replied:
I'd take a different approach:
- leave the seed task as it is (perhaps include the task to fill the languages)
- include this rake task in a cap task that executes each time a deploy happens
then all bases are covered.

lotophage replied:
The seed task already includes the task to fill in the languages. If the cap task executes each time deploy happens, but the first thing these task do is call `Role.delete_all` and `Language.delete_all`, but that won't rest the id's in the database will it? What would be the consequence of ActiveRecord associations for existing users with roles for example?

dyazbek-intersect replied:
I think we are talking about the same thing here - Integrating it in to the seed task is good, but the seed task is only run once (at first deploy) i.e. not at each redeploy, so what I'm trying to say is that you'll need to call populate_languages elsewhere too - somewhere that gets run each deploy otherwise the task won't get executed next time you deploy to prod.

shaun-intersect replied:
So currently this 'populate_languages' method is defined only within the seed helper, rather than a standalone rake task. Can somewhere else call this method from the seed helper? If so where should it be called from?

dyazbek-intersect replied:
I'm not sure - you'll need to work out what is called for each deploy and integrate it into there...

shaun-intersect replied:
I've added a hook to deployment to run the method to populate the languages table from the CSV. I think this should be ok, but I don't have any method for testing it.

lotophage commented:
Looks okay to me. I think you should merge and deploy to QA unless Dan has any more comments?

shaun-intersect commented:
I've made a few small additional changes since the last review, to do with remembering form input and flagging that the item needs to be reindexed before it can be viewed.

lotophage commented on app/controllers/collections_controller.rb:
```diff
@@ -504,11 +525,18 @@ def update_document_source_in_graph(jsonld_graph, doc_identifier, doc_source)
+  #
+  def format_document_source_metadata(doc_source)
+    # Escape any filename spaces with '%20' as URIs with spaces are flagged as invalid when RDF loads
+    {'@id' => "file://#{doc_source.sub(" ", "%20")}"}
```
Okay for now, but I think that Rails has it's own escape methods so you don't need to reinvent the wheel. There are probably other cases that also need to be escaped, such as having an ampersand in the file name.

lotophage commented:
Looks okay, merge at will
